File: sdk/python/src/stellarmesh_logging/logger.py
"""Logger facade backed by an explicitly configured client."""

from __future__ import annotations

import asyncio
import sys
import threading
import traceback
from datetime import UTC, datetime
from typing import Any

from .client import Client
from .contracts import Level


class Logger:
    """Module-scoped structured logger."""

    def __init__(
        self,
        name: str,
        *,
        client: Client,
        service: str | None = None,
        bound_metadata: dict[str, Any] | None = None,
    ) -> None:
        self.name = name
        self._client = client
        self._service = service or client.config.service
        self._bound_metadata = bound_metadata or {}

    def bind(self, **metadata: Any) -> Logger:
        """Return a logger with metadata attached to every event."""
        return Logger(
            self.name,
            client=self._client,
            service=self._service,
            bound_metadata={**self._bound_metadata, **metadata},
        )

    def debug(
        self, message: str, *, trace_id: str | None = None, **metadata: Any
    ) -> bool:
        return self._emit(Level.DEBUG, message, trace_id, metadata)

    def info(
        self, message: str, *, trace_id: str | None = None, **metadata: Any
    ) -> bool:
        return self._emit(Level.INFO, message, trace_id, metadata)

    def warning(
        self, message: str, *, trace_id: str | None = None, **metadata: Any
    ) -> bool:
        return self._emit(Level.WARNING, message, trace_id, metadata)

    def error(
        self, message: str, *, trace_id: str | None = None, **metadata: Any
    ) -> bool:
        return self._emit(Level.ERROR, message, trace_id, metadata)

    def audit(
        self, message: str, *, trace_id: str | None = None, **metadata: Any
    ) -> bool:
        return self._emit(Level.AUDIT, message, trace_id, metadata)

    def exception(
        self, message: str, *, trace_id: str | None = None, **metadata: Any
    ) -> bool:
        exc_type, exc_value, exc_traceback = sys.exc_info()
        if exc_type is not None:
            metadata.setdefault("exception_type", exc_type.__name__)
            metadata.setdefault("exception_message", str(exc_value))
            metadata.setdefault(
                "traceback",
                "".join(traceback.format_exception(exc_type, exc_value, exc_traceback)),
            )
        return self._emit(Level.ERROR, message, trace_id, metadata)

    def _emit(
        self,
        level: Level,
        message: str,
        trace_id: str | None,
        metadata: dict[str, Any],
    ) -> bool:
        return self._client.emit_event(
            level,
            message=message,
            trace_id=trace_id,
            service=self._service,
            timestamp=datetime.now(UTC),
            metadata={"logger": self.name, **self._bound_metadata, **metadata},
        )


_default_client: Client | None = None
_default_client_lock = threading.Lock()


def set_default_client(client: Client | None) -> None:
    """Replace the process-wide client without hiding its configuration."""
    global _default_client
    with _default_client_lock:
        previous = _default_client
        _default_client = client
    if previous is not None and previous is not client:
        previous.close()


def get_logger(
    name: str, *, client: Client | None = None, service: str | None = None
) -> Logger:
    """Create a logger from an explicit or previously configured default client."""
    resolved = client
    if resolved is None:
        with _default_client_lock:
            resolved = _default_client
    if resolved is None:
        raise RuntimeError("no default Stellarmesh logging client is configured")
    return Logger(name, client=resolved, service=service)


async def shutdown_logging(*, timeout: float = 2.0) -> bool:
    """Detach and asynchronously drain the default client."""
    global _default_client
    with _default_client_lock:
        client = _default_client
        _default_client = None
    if client is None:
        return True
    return await client.aclose(timeout=timeout)


def shutdown_logging_sync(*, timeout: float = 2.0) -> bool:
    """Synchronously drain the default client for non-async entrypoints."""
    try:
        asyncio.get_running_loop()
    except RuntimeError:
        return asyncio.run(shutdown_logging(timeout=timeout))
    raise RuntimeError("use await shutdown_logging() from an active event loop")

File: sdk/go/logging/logger.go
package logging

import (
	"context"
	"errors"
	"strings"
	"time"
)

// Emitter 接收一个事件，不阻塞等待远端投递。
type Emitter interface {
	Emit(context.Context, Event) bool
}

// TraceIDProvider 从调用方管理的上下文状态中解析 trace id。
type TraceIDProvider func(context.Context) string

// LoggerConfig 配置面向应用的日志记录器。
type LoggerConfig struct {
	Service         string
	Emitter         Emitter
	Now             func() time.Time
	TraceIDProvider TraceIDProvider
	MinimumLevel    Level
}

// Logger 为单个服务构造结构化事件。
type Logger struct {
	service         string
	emitter         Emitter
	now             func() time.Time
	traceIDProvider TraceIDProvider
	minimumLevel    Level
}

// NewLogger 创建结构化日志门面。
func NewLogger(config LoggerConfig) (*Logger, error) {
	if strings.TrimSpace(config.Service) == "" || config.Service != strings.TrimSpace(config.Service) {
		return nil, errors.New("logging service name must be non-empty and trimmed")
	}
	if config.Emitter == nil {
		return nil, errors.New("logging emitter is required")
	}
	now := config.Now
	if now == nil {
		now = time.Now
	}
	minimumLevel := config.MinimumLevel
	if minimumLevel == "" {
		minimumLevel = LevelDebug
	}
	if err := minimumLevel.Validate(); err != nil {
		return nil, err
	}
	return &Logger{
		service: config.Service, emitter: config.Emitter, now: now,
		traceIDProvider: config.TraceIDProvider, minimumLevel: minimumLevel,
	}, nil
}

func (logger *Logger) Debug(ctx context.Context, message, traceID string, metadata map[string]any) bool {
	return logger.emit(ctx, LevelDebug, message, traceID, metadata)
}

func (logger *Logger) Info(ctx context.Context, message, traceID string, metadata map[string]any) bool {
	return logger.emit(ctx, LevelInfo, message, traceID, metadata)
}

func (logger *Logger) Warning(ctx context.Context, message, traceID string, metadata map[string]any) bool {
	return logger.emit(ctx, LevelWarning, message, traceID, metadata)
}

func (logger *Logger) Error(ctx context.Context, message, traceID string, metadata map[string]any) bool {
	return logger.emit(ctx, LevelError, message, traceID, metadata)
}

func (logger *Logger) Audit(ctx context.Context, message, traceID string, metadata map[string]any) bool {
	return logger.emit(ctx, LevelAudit, message, traceID, metadata)
}

func (logger *Logger) emit(ctx context.Context, level Level, message, traceID string, metadata map[string]any) bool {
	if logger.emitter == nil {
		return false
	}
	if !levelEnabled(level, logger.minimumLevel) {
		return false
	}
	if traceID == "" && logger.traceIDProvider != nil {
		traceID = logger.traceIDProvider(ctx)
	}
	return logger.emitter.Emit(ctx, Event{
		Timestamp: logger.now().UTC(),
		Level:     level, Service: logger.service, Message: message, TraceID: traceID,
		Metadata: SanitizeMetadata(metadata),
	})
}

func levelEnabled(level, minimum Level) bool {
	return levelOrder(level) >= levelOrder(minimum)
}

func levelOrder(level Level) int {
	switch level {
	case LevelDebug:
		return 0
	case LevelInfo:
		return 1
	case LevelWarning:
		return 2
	case LevelError:
		return 3
	case LevelAudit:
		return 4
	default:
		return -1
	}
}
